### DISCUSSION TITLE
JSON Web Key Secret URI?

### DISCUSSION THREAD

**Discussion by zvictor** [ory/kratos#2302] [Q&A] [answered]
Hello everyone!

Sorry in advance for the dumb question (security is not my expertise), but I have been looking for an IdP that I can use to support [Faunadb's security model](https://docs.fauna.com/fauna/current/security/).
They offer [integration with external identity providers](https://fauna.com/blog/choosing-an-authentication-strategy-with-fauna) and native [IdP-agnostic ABAC](https://docs.fauna.com/fauna/current/security/external/).

There are plenty of material on [how to connect faunadb and auth0](https://fauna.com/blog/setting-up-sso-authentication-in-fauna-with-auth0) 🤢 but nothing on Ory/Kratos.

# Can I use Ory's solutions as the IdP for fauna?

I tried figuring it out by myself, but after setting up Kratos I got stuck missing a `JSON Web Key Secret URI`.
AFAIK only Hydra works with Jwks, not Kratos, but I am not confident that that's what I should be going for.
My limited understanding on this topic makes me suspicious that I am overcomplicating things....

**Is Hydra the way to go here?
Can it provide the right interface to be the IdP of Fauna or am I missing something?**

**Comment by vinckr** [comment, marked as answer]
Hello @zvictor 
Would love to support you with creating a guide how to set this up with Ory. 

It seems fauna db requires a OAuth2 based login method, so Ory Hydra would probably be the way to go. 
Since Ory Kratos does not use OAuth2 you might need both. I am not entirely sure.
Have a look at this article that contrasts the Auth0 flow with an Ory Kratos / Ory Hydra based flow. 
Keep in mind that the integration is not super easy right now, but there is work underway to make it plug'n'play - will be available in Q2. 

Am I right in the assumption that this will be mostly API based access and not involving frontend UIs for connecting to fauna?

**Reply by zvictor** [reply]
Hello @vinckr!
Thank you for the thoughtful answer.

Let me give you a bit more context:
* I would be super happy to use Ory Cloud instead, but it does not support social login yet and I need it up and running (with Github login) in a month from now.
* I need this for a new project of mine, but I'd also love to share this integration as an easy-to-clone example in the tool I have been building (https://zvictor.github.io/faugra). I have already built 2 other [authentication models](https://github.com/zvictor/faugra/tree/master/examples) using fauna + nextjs + iron-session there.


---

> Am I right in the assumption that this will be mostly API based access and not involving frontend UIs for connecting to fauna?

I expect Kratos/Hydra to provide a token/`ory_kratos_session` that can then be sent to fauna, either from frontend or backend, to be used to authorize and give access to documents owned by that user.
The user will not be aware in any level that Faunadb exists, so there should be no frontend UIs for that. Not sure if that's what you asked...

> Have a look at this article that contrasts the Auth0 flow with an Ory Kratos / Ory Hydra based flow.

I think you forgot the link there.


> Would love to support you with creating a guide how to set this up with Ory.

That's something I would be interested in, and I think that @ptpaterson and @rts-rob would be as well!
They could pay [500$ for a post like that](https://fauna.com/blog/write-with-fauna)! 😊🤌

**Reply by vinckr** [reply]
Woops forgot the link 😅 
https://blog.px.dev/open-source-auth/

**Comment by zepatrik** [comment]
One other option would also be to use Oathkeeper (or something similar) to exchange Kratos sessions for a JWT, basically putting it between the user and faunadb. From a first glance that should work.

**Reply by zvictor** [reply]
Interesting observation @zepatrik!
After getting started with Oathkeeper, what should I be looking at in order to go into the right direction ?

**Reply by vinckr** [reply]
Hey Victor, 
this is the functionality that Patrik mentioned: 
https://www.ory.sh/docs/oathkeeper/pipeline/mutator#id_token

You can then use the jwks endpoint from Oathkeeper 👍 .
